packages/core/src/Room.ts

```diff
@@ -583,62 +587,73 @@ export abstract class Room<State extends object= any, Metadata= any> {
+    previousClient: Client,
+    seconds: number | 'manual',
+    abortController?: AbortController,
+  ): Promise<void> {
     if (this.internalState === RoomInternalState.DISCONNECTING) {
       this._disposeIfEmpty(); // gracefully shutting down
       throw new Error('disconnecting');
     }
 
-    const sessionId = previousClient.sessionId;
-    const reconnectionToken = previousClient._reconnectionToken;
+    return new Promise(async (resolve, reject) => {
+      if (seconds === undefined) { // TODO: remove this check
+        console.warn('DEPRECATED: allowReconnection() requires a second argument. Using "manual" mode.');
+        seconds = 'manual';
+      }
+
+      if (seconds === 'manual') {
+        seconds = Infinity;
+      }
 
-    this._reserveSeat(sessionId, true, seconds, true);
+      const sessionId = previousClient.sessionId;
+      const reconnectionToken = previousClient._reconnectionToken;
 
-    // keep reconnection reference in case the user reconnects into this room.
-    const reconnection = new Deferred<Client>();
-    this._reconnections[reconnectionToken] = [sessionId, reconnection];
+      await this._reserveSeat(sessionId, true, seconds, true);
 
-    if (seconds !== Infinity) {
-      // expire seat reservation after timeout
-      this.reservedSeatTimeouts[sessionId] = setTimeout(() =>
-        reconnection.reject(false), seconds * 1000);
-    }
+      // keep reconnection reference in case the user reconnects into this room.
+      const reconnection = new Deferred<Client>();
+      this._reconnections[reconnectionToken] = [sessionId, reconnection];
 
-    const cleanup = () => {
-      delete this._reconnections[reconnectionToken];
-      delete this.reservedSeats[sessionId];
-      delete this.reservedSeatTimeouts[sessionId];
-      this._reconnectingSessionId.delete(sessionId);
-    };
-
-    reconnection.
-      then((newClient) => {
-        newClient.auth = previousClient.auth;
-        newClient.userData = previousClient.userData;
-        previousClient.ref = newClient.ref; // swap "ref" for convenience
-        previousClient.state = ClientState.RECONNECTED;
-        clearTimeout(this.reservedSeatTimeouts[sessionId]);
-        cleanup();
-      }).
-      catch(() => {
-        cleanup();
-        this.resetAutoDisposeTimeout();
-      });
+      if (seconds !== Infinity) {
+        // expire seat reservation after timeout
+        this.reservedSeatTimeouts[sessionId] = setTimeout(() =>
+          reconnection.reject(false), seconds * 1000);
+      }
+
+      if (abortController) {
+        abortController.signal.addEventListener('abort', () =>
+          reconnection.reject(false));
+      }
 
-    return reconnection;
+      const cleanup = () => {
+        delete this._reconnections[reconnectionToken];
+        delete this.reservedSeats[sessionId];
+        delete this.reservedSeatTimeouts[sessionId];
+        this._reconnectingSessionId.delete(sessionId);
+      };
+
+      reconnection
```

Comment:
I've never been able to get the reconnection aspect to work well with colyseus so I'm not sure of the impact here, but it should still RETURN something. This has been change to have no return value, so the call cannot be awaited anymore